### PR TITLE
PREAPPS-5503: adding the unreadDescendent to check if any subfolder or the folder itself has a unread mails in the hireracy

### PR DESCRIPTION
**Pull request by auminpatel**
Function setUnreadDescendentFlag will check only if the folder View is null or Message , null because the follders that are through imap or pop
hasUnreadDescendent will set the flag unreadDescendent if a folders subfolder or itself has a unread count

### CODE REVIEW

File: src/batch-client/index.ts [hunk 1]
```diff
@@ -169,6 +169,49 @@ function normalizeMessage(
 	);
 }
 
+const hasUnreadDescendent = (folder: any) => {
+	if (
+		folder[
+			folder.name === 'Drafts' || folder.name === 'Outbox'
```
Comment (silentsakky) [inline src/batch-client/index.ts:175]: instead of checking folder names, can we check folder ids?

File: src/batch-client/index.ts [hunk 2]
```diff
@@ -169,6 +169,49 @@ function normalizeMessage(
...
+				? 'nonFolderItemCount'
+				: 'unread'
+		] > 0
+	) {
+		return true;
+	}
+
+	const folderArray = get(folder, 'folders') || [];
+	for (let i = 0; i < folderArray.length; i++) {
```
Comment (silentsakky) [inline src/batch-client/index.ts:184]: one small optimization to not count length with every iteration
`for (let i = 0, len = folderArray.length; i < len; i++) {`

File: src/batch-client/index.ts [hunk 3]
```diff
@@ -169,6 +169,49 @@ function normalizeMessage(
...
+		if (hasUnreadDescendent(folderArray[i])) {
+			return true;
+		}
```
Comment (silentsakky) [inline src/batch-client/index.ts:187]: `return hasUnreadDescendent(folderArray[i]);`

File: src/batch-client/index.ts [hunk 4]
```diff
@@ -169,6 +169,49 @@ function normalizeMessage(
...
+	if (!view || view === FolderView.Message) {
+		folder = {
+			...folder,
+			unreadDescendent: hasUnreadDescendent(folder)
+		};
+
+		if (folderArray) {
+			folder = {
+				...folder,
+				folders: folderArray.map(setUnreadDescendentFlag)
```
Comment (silentsakky) [inline src/batch-client/index.ts:208]: hasUnreadDescendent is already doing a loop on subfolders, here we are also doing same.
can we reduce looping required to achieve end result?

File: src/batch-client/index.ts (hunk 1 above)
Comment (auminpatel) [reply src/batch-client/index.ts:175]: I couldnt find a folderId constant for Outbox so wll keep this, name only

File: src/batch-client/index.ts (hunk 2 above)
Comment (auminpatel) [reply src/batch-client/index.ts:184]: Done

File: src/batch-client/index.ts (hunk 3 above)
Comment (auminpatel) [reply src/batch-client/index.ts]: Done

File: src/batch-client/index.ts (hunk 4 above)
Comment (auminpatel) [reply src/batch-client/index.ts:211]: Have tried,but couldnt find a solution ,to reduce looping , for certian level when a unreadDescendent is laready present it wont loop over , but we do need to check into deeper level 
https://github.com/Zimbra/zm-api-js-client/pull/447/files#diff-2f62190d7ad815da92e5dcf270fdd6e290f3362cf453d2951e2f4d7ed46e2badR182-R183
t

**Review by silentsakky — APPROVED** [review]

**Review by lavegupta — APPROVED** [review]

**Pull request merged**